File: Baekjoon/graphTraversal/1303.cpp
#include <iostream>
#include <vector>
#include <queue>
#define MAX 101

using namespace std;

int N, M;
char soldiers[MAX][MAX];
bool visited[MAX][MAX];
int dx[4] = { -1, 1, 0, 0 };
int dy[4] = { 0, 0, -1, 1 };

int bfs(int i, int j, char c, int N, int M) {
	queue<pair<int, int>> q;
	q.push({ i, j });
	visited[i][j] = 1;
	int cnt = 1;

	while (!q.empty()){
		int x = q.front().first;
		int y = q.front().second;
		q.pop();

		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];
			if (nx >= 0 && nx < M && ny >= 0 && ny < N){
				if (visited[nx][ny] == 0 && soldiers[nx][ny] == c) {
					visited[nx][ny] = 1;
					q.push({ nx, ny });
					cnt++;
				}
			}
		} 
	}
	return cnt;
}

int main() {
	int N, M;
	cin >> N >> M;
	for (int i = 0; i < M; i++) {
		string tmp;
		cin >> tmp;
		for (int j = 0; j < N; j++) {
			soldiers[i][j] = tmp[j];
		}
	}

	int W = 0;
	int B = 0;
	for (int i = 0; i < M; i++) {
		for (int j = 0; j < N; j++) {
			if (visited[i][j] == 0) {
				int value = bfs(i, j, soldiers[i][j], N , M);
				if (soldiers[i][j] == 'W') {
					W += value * value;
				}
					
				else {
					B += value * value;
				}
			}
		}
	}

	cout << W << ' ' << B;
}

File: Baekjoon/graphTraversal/1303.py
from sys import stdin
from collections import deque
N, M = map(int, stdin.readline().split())
soldiers = [list(stdin.readline().rstrip()) for _ in range(M)]
visited = [[0] * M for _ in range(N)]
queue = deque()

def bfs(x, y, c):
    queue.append((x, y))
    visited[x][y] = 1
    cnt = 1

    while queue:
        x, y = queue.popleft()

        for dx, dy in (1,0), (-1,0), (0, 1), (0, -1):
            nx, ny = x + dx, y + dy

            if nx < 0 or nx >= M or ny < 0 or ny >= N:
                continue

            if soldiers[nx][ny] != c or visited[nx][ny] != 0:
                continue

            visited[nx][ny] = 1
            queue.append((nx, ny))
            cnt = cnt + 1
    return cnt

W_sum = 0
B_sum = 0
for i in range(N):
    for j in range(M):
        if visited[i][j] == 0:
            value = bfs(i, j, soldiers[i][j])
            if soldiers[i][j] == 'W':
                W_sum += value ** 2
            else:
                B_sum += value ** 2
print(W_sum, B_sum)